{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook generates the .mut file needed for ddG calculations using Triad."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import product\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9261"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/disk2/fli/SSMuLA/data/DHFR/processed/DHFR.csv\")\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9261"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df[\"AAs\"].values.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the GB1_Fitness file and build mutants using it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load GB1 fitness\n",
    "fitness_df = pd.read_csv(\"/disk1/jyang4/repos/data/Pgb_fitness.csv\")\n",
    "measured = fitness_df.Combo.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Combo</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Alignment Probability</th>\n",
       "      <th>Alignment Count</th>\n",
       "      <th>Column</th>\n",
       "      <th>StdArea</th>\n",
       "      <th>Iso1Area</th>\n",
       "      <th>Iso2Area</th>\n",
       "      <th>num_wells</th>\n",
       "      <th>...</th>\n",
       "      <th>EVMutation Rank</th>\n",
       "      <th>Triad Score</th>\n",
       "      <th>Triad Rank</th>\n",
       "      <th>NormIso1_recomb</th>\n",
       "      <th>NormIso2_recomb</th>\n",
       "      <th>NormIso1_recomb_rank</th>\n",
       "      <th>NormIso2_recomb_rank</th>\n",
       "      <th>NormIso1</th>\n",
       "      <th>NormIso2</th>\n",
       "      <th>Diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FRMNY</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.624151</td>\n",
       "      <td>339.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>484.0</td>\n",
       "      <td>3398.0</td>\n",
       "      <td>567.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>69.0</td>\n",
       "      <td>-648.97722</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.586441</td>\n",
       "      <td>0.765307</td>\n",
       "      <td>3.821134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SAFRY</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.755508</td>\n",
       "      <td>128.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>3778.0</td>\n",
       "      <td>963.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>118.0</td>\n",
       "      <td>-646.67160</td>\n",
       "      <td>48</td>\n",
       "      <td>0.835324</td>\n",
       "      <td>1.329356</td>\n",
       "      <td>20.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>5.067932</td>\n",
       "      <td>1.291800</td>\n",
       "      <td>3.776133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GIDLY</td>\n",
       "      <td>165.0</td>\n",
       "      <td>0.086459</td>\n",
       "      <td>197.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>2485.0</td>\n",
       "      <td>428.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>112.0</td>\n",
       "      <td>-642.32678</td>\n",
       "      <td>97</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>155.5</td>\n",
       "      <td>155.5</td>\n",
       "      <td>3.449043</td>\n",
       "      <td>0.594040</td>\n",
       "      <td>2.855002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TSGMY</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.244922</td>\n",
       "      <td>154.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>497.0</td>\n",
       "      <td>3058.0</td>\n",
       "      <td>1062.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>90.0</td>\n",
       "      <td>-645.07324</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.019563</td>\n",
       "      <td>1.395937</td>\n",
       "      <td>2.623626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TNMPY</td>\n",
       "      <td>331.0</td>\n",
       "      <td>0.867393</td>\n",
       "      <td>241.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>484.0</td>\n",
       "      <td>1788.0</td>\n",
       "      <td>367.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>146.0</td>\n",
       "      <td>-633.74115</td>\n",
       "      <td>162</td>\n",
       "      <td>0.262932</td>\n",
       "      <td>0.495441</td>\n",
       "      <td>52.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>3.260248</td>\n",
       "      <td>0.669190</td>\n",
       "      <td>2.591058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>183</td>\n",
       "      <td>LQSGA</td>\n",
       "      <td>353.0</td>\n",
       "      <td>0.172635</td>\n",
       "      <td>238.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>474.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>2725.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>130.0</td>\n",
       "      <td>-642.98517</td>\n",
       "      <td>84</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>155.5</td>\n",
       "      <td>155.5</td>\n",
       "      <td>0.513877</td>\n",
       "      <td>5.073604</td>\n",
       "      <td>-4.559727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>186</td>\n",
       "      <td>QYKGD</td>\n",
       "      <td>286.0</td>\n",
       "      <td>0.001800</td>\n",
       "      <td>177.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>442.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>80.0</td>\n",
       "      <td>-641.25164</td>\n",
       "      <td>113</td>\n",
       "      <td>0.390436</td>\n",
       "      <td>4.028886</td>\n",
       "      <td>38.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.213644</td>\n",
       "      <td>4.863887</td>\n",
       "      <td>-4.650243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>213</td>\n",
       "      <td>PCLTD</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0.717479</td>\n",
       "      <td>513.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>506.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>3513.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>216.0</td>\n",
       "      <td>-629.35627</td>\n",
       "      <td>187</td>\n",
       "      <td>0.075245</td>\n",
       "      <td>2.350554</td>\n",
       "      <td>87.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.120283</td>\n",
       "      <td>4.801761</td>\n",
       "      <td>-4.681477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>190</td>\n",
       "      <td>NNVER</td>\n",
       "      <td>306.0</td>\n",
       "      <td>0.687294</td>\n",
       "      <td>230.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>2847.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>176.0</td>\n",
       "      <td>-640.64290</td>\n",
       "      <td>120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.183788</td>\n",
       "      <td>4.936261</td>\n",
       "      <td>-4.752473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>215</td>\n",
       "      <td>TSGRE</td>\n",
       "      <td>193.0</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>556.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>3496.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>185.0</td>\n",
       "      <td>-639.24294</td>\n",
       "      <td>135</td>\n",
       "      <td>0.093254</td>\n",
       "      <td>1.108917</td>\n",
       "      <td>79.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.151029</td>\n",
       "      <td>4.934558</td>\n",
       "      <td>-4.783529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0.1  Combo  Unnamed: 0  Alignment Probability  Alignment Count  \\\n",
       "0               0  FRMNY         7.0               0.624151            339.0   \n",
       "1               1  SAFRY         2.0               0.755508            128.0   \n",
       "2               2  GIDLY       165.0               0.086459            197.0   \n",
       "3               3  TSGMY        28.0               0.244922            154.0   \n",
       "4               5  TNMPY       331.0               0.867393            241.0   \n",
       "..            ...    ...         ...                    ...              ...   \n",
       "211           183  LQSGA       353.0               0.172635            238.0   \n",
       "212           186  QYKGD       286.0               0.001800            177.0   \n",
       "213           213  PCLTD       251.0               0.717479            513.0   \n",
       "214           190  NNVER       306.0               0.687294            230.0   \n",
       "215           215  TSGRE       193.0               0.001915            556.0   \n",
       "\n",
       "     Column  StdArea  Iso1Area  Iso2Area  num_wells  ...  EVMutation Rank  \\\n",
       "0       2.0    484.0    3398.0     567.0          1  ...             69.0   \n",
       "1       8.0    487.0    3778.0     963.0          1  ...            118.0   \n",
       "2       5.0    501.0    2485.0     428.0          1  ...            112.0   \n",
       "3       3.0    497.0    3058.0    1062.0          1  ...             90.0   \n",
       "4      11.0    484.0    1788.0     367.0          1  ...            146.0   \n",
       "..      ...      ...       ...       ...        ...  ...              ...   \n",
       "211    12.0    474.0     276.0    2725.0          1  ...            130.0   \n",
       "212     4.0    442.0     107.0    2436.0          1  ...             80.0   \n",
       "213     1.0    506.0      88.0    3513.0          1  ...            216.0   \n",
       "214     7.0    509.0     106.0    2847.0          1  ...            176.0   \n",
       "215     2.0    490.0     107.0    3496.0          1  ...            185.0   \n",
       "\n",
       "     Triad Score  Triad Rank  NormIso1_recomb  NormIso2_recomb  \\\n",
       "0     -648.97722          33              NaN              NaN   \n",
       "1     -646.67160          48         0.835324         1.329356   \n",
       "2     -642.32678          97         0.000000         0.000000   \n",
       "3     -645.07324          60              NaN              NaN   \n",
       "4     -633.74115         162         0.262932         0.495441   \n",
       "..           ...         ...              ...              ...   \n",
       "211   -642.98517          84         0.000000         0.000000   \n",
       "212   -641.25164         113         0.390436         4.028886   \n",
       "213   -629.35627         187         0.075245         2.350554   \n",
       "214   -640.64290         120              NaN              NaN   \n",
       "215   -639.24294         135         0.093254         1.108917   \n",
       "\n",
       "     NormIso1_recomb_rank  NormIso2_recomb_rank  NormIso1  NormIso2      Diff  \n",
       "0                     NaN                   NaN  4.586441  0.765307  3.821134  \n",
       "1                    20.0                  70.0  5.067932  1.291800  3.776133  \n",
       "2                   155.5                 155.5  3.449043  0.594040  2.855002  \n",
       "3                     NaN                   NaN  4.019563  1.395937  2.623626  \n",
       "4                    52.0                 132.0  3.260248  0.669190  2.591058  \n",
       "..                    ...                   ...       ...       ...       ...  \n",
       "211                 155.5                 155.5  0.513877  5.073604 -4.559727  \n",
       "212                  38.0                   2.0  0.213644  4.863887 -4.650243  \n",
       "213                  87.0                  19.0  0.120283  4.801761 -4.681477  \n",
       "214                   NaN                   NaN  0.183788  4.936261 -4.752473  \n",
       "215                  79.0                  83.0  0.151029  4.934558 -4.783529  \n",
       "\n",
       "[216 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0.1', 'Combo', 'Unnamed: 0', 'Alignment Probability',\n",
       "       'Alignment Count', 'Column', 'StdArea', 'Iso1Area', 'Iso2Area',\n",
       "       'num_wells', 'EVMutation', 'EVMutation Rank', 'Triad Score',\n",
       "       'Triad Rank', 'NormIso1_recomb', 'NormIso2_recomb',\n",
       "       'NormIso1_recomb_rank', 'NormIso2_recomb_rank', 'NormIso1', 'NormIso2',\n",
       "       'Diff'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "AAs = ('A','R','N','D','C','Q','E','G','H','I','L','K','M','F','P','S','T','W','Y','V')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "all = [''.join(x) for x in list(product(AAs, repeat=5))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#variants = list(set(all) - set(measured))\n",
    "variants = set(measured)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the full mut file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the wildtype amino acids at each position\n",
    "wt_aas = (\"W\", \"Y\", \"W\", \"V\", \"F\")\n",
    "\n",
    "# Define the prefixes for mutagenesis\n",
    "prefixes = (\"A_56\", \"A_57\", \"A_59\", \"A_60\", \"A_89\")\n",
    "\n",
    "# Get the variants\n",
    "#all_variants = list(product(AAs, repeat=4))\n",
    "all_variants = variants\n",
    "\n",
    "# Loop over variants\n",
    "mutation_encodings = []\n",
    "for variant in all_variants:\n",
    "\n",
    "    # Loop over each character in the variant\n",
    "    mut_encoding_list = []\n",
    "    for j, (var_char, wt_char) in enumerate(zip(variant, wt_aas)):\n",
    "\n",
    "        # If the var_char does not equal the wt_char, append\n",
    "        if var_char != wt_char:\n",
    "            mut_encoding_list.append(prefixes[j] + var_char)\n",
    "            \n",
    "    # If the mut_encoding_list has no entries, continue (this is wild type)\n",
    "    if len(mut_encoding_list) == 0:\n",
    "        continue\n",
    "    \n",
    "    # Otherwise, append to mutation_encodings\n",
    "    else:\n",
    "        mutation_encodings.append(\"+\".join(mut_encoding_list) + \"\\n\")    \n",
    "        \n",
    "# Save the mutants\n",
    "with open(\"ParPgb_5site_round1.mut\", \"w\") as f:\n",
    "    f.writelines(mutation_encodings)\n",
    "    \n",
    "# # Save the mutants\n",
    "# with open(\"2GI9_Test.mut\", \"w\") as f:\n",
    "#     f.writelines(mutation_encodings[:400])    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sanity check to make sure the mutation_encodings map back to the mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assert that we have the correct number of mutations\n",
    "assert len(mutation_encodings) == 149360\n",
    "\n",
    "# Compile a regular expression for finding variants\n",
    "position_re = re.compile(\"_([0-9]+)([A-Z]{1})\")\n",
    "\n",
    "# Pull the variants\n",
    "variants = fitness_df.Mutations.values.tolist()[1:]\n",
    "\n",
    "# Define a dicitonary linking back to wt aa\n",
    "wt_aa_dict = {\"39\": \"V\",\n",
    "             \"40\": \"D\",\n",
    "             \"41\": \"G\",\n",
    "             \"54\": \"V\"}\n",
    "\n",
    "# Reconstruct variants\n",
    "for mutation_encoding, variant in zip(mutation_encodings, variants):\n",
    "    \n",
    "    # Find all matches\n",
    "    all_matches = position_re.findall(mutation_encoding)\n",
    "    \n",
    "    # Reconstruct the variant and make sure it's what we expect\n",
    "    reconstructed_variant = \"\".join([f\"{wt_aa_dict[pos]}{pos}{new_char}\" for pos, new_char in all_matches])\n",
    "    assert reconstructed_variant == variant"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 ('triad')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "91758922c0a205f7645a835882268f67fbb45058931ebca30bd11a16ce900e1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
